lib/spack/spack/environment.py

```
@@ -540,36 +594,48 @@ def regenerate(self, all_specs, roots):
+            # we always regenerate the view from scratch.
+            # We will do this by hashing the view contents and putting the view
+            # in a directory by hash, and then having a symlink to the real
+            # view in the root. The real root for a view at /dirname/basename
+            # will be /dirname/._basename_<hash>.
+            # This allows for atomic swaps when we update the view
+
+            # cache the roots because the way we determine which is which does
+            # not work while we are updating
+            new_root = self._next_root(installed_specs_for_view)
+            old_root = self._current_root
+
+            if new_root == old_root:
+                tty.msg("View at %s does not need regeneration." % self.root)
+                return
+
+            # construct view at new_root
+            tty.msg("Updating view at {0}".format(self.root))
+
+            view = self.view(new=new_root)
+            fs.mkdirp(new_root)
+            view.add_specs(*installed_specs_for_view,
+                           with_dependencies=False)
+
+            # create symlink from tmpname to new_root
+            root_dirname = os.path.dirname(self.root)
+            tmp_symlink_name = os.path.join(root_dirname, '._view_link')
+            os.symlink(new_root, tmp_symlink_name)
+
+            # mv symlink atomically over root symlink to old_root
+            if os.path.exists(self.root) and not os.path.islink(self.root):
+                msg = "Cannot create view: "
+                msg += "file already exists and is not a link: %s" % self.root
+                raise SpackEnvironmentViewError(msg)
+            os.rename(tmp_symlink_name, self.root)
+
+            # remove old_root
+            if old_root and os.path.exists(old_root):
+                try:
+                    shutil.rmtree(old_root)
+                except (IOError, OSError):
+                    tty.warn("Failed to remove old view at %s" % old_root)
```

this should print the error -- currently it swallows it.  Probably you also want something like `if spack.debug: raise` in here.

---

We shouldn't raise here on debug -- the user may care about the generation of the other views, which this would prevent. I've added the error message to the warning instead.